crates/hir-ty/src/chalk_db.rs

```diff
@@ -884,13 +927,378 @@ fn impl_def_datum(db: &dyn HirDatabase, krate: Crate, impl_id: hir_def::ImplId)
     Arc::new(impl_datum)
 }
 
+fn impl_rpitit_values(
+    db: &dyn HirDatabase,
+    impl_id: hir_def::ImplId,
+    impl_trait: &Binders<TraitRef>,
+    impl_items: &ImplItems,
+    trait_datum: &TraitDatum,
+) -> impl Iterator<Item = AssociatedTyValueId> {
+    let mut trait_rpitit_to_impl_rpitit = FxHashMap::default();
+
+    return trait_datum
+        .associated_ty_ids
+        .iter()
+        .filter_map(|&trait_assoc| match from_assoc_type_id(db, trait_assoc) {
+            AnyTraitAssocType::Rpitit(it) => Some(it),
+            AnyTraitAssocType::Normal(_) => None,
+        })
+        .map(move |trait_assoc_id| {
+            if let Some(&impl_assoc_id) = trait_rpitit_to_impl_rpitit.get(&trait_assoc_id) {
+                return impl_assoc_id;
+            }
+
+            let trait_assoc = trait_assoc_id.loc(db);
+            let trait_method_id = trait_assoc.synthesized_from_method;
+            let trait_method_generics = generics(db.upcast(), trait_method_id.into());
+
+            let impl_assoc_id = (|| {
```

Comment:
what's the immediately called closure for? anything that couldn't be done with `'foo: { break 'foo value; }`?

Reply:
It could, that's just the idiom until we have try blocks.

Reply:
Hm, did I miss where you use the ? operator? 😅